File: vsg_qt/job_queue_dialog/ui.py
# vsg_qt/job_queue_dialog/ui.py
from __future__ import annotations

from collections.abc import Callable

from PySide6.QtCore import QItemSelectionModel, Qt
from PySide6.QtGui import QKeySequence, QShortcut
from PySide6.QtWidgets import (
    QAbstractItemView,
    QDialog,
    QDialogButtonBox,
    QHBoxLayout,
    QMenu,
    QPushButton,
    QTableWidget,
    QVBoxLayout,
)

from vsg_qt.add_job_dialog import AddJobDialog

from .logic import JobQueueLogic


class JobQueueDialog(QDialog):
    def __init__(
        self,
        config: AppConfig,
        log_callback: Callable[[str], None],
        layout_manager: JobLayoutManager,
        parent=None,
    ):
        super().__init__(parent)
        self.setWindowTitle("Job Queue")
        self.setMinimumSize(1200, 600)

        self.config = config
        self.log_callback = log_callback
        self._logic = JobQueueLogic(self, layout_manager)

        self._build_ui()
        self._connect_signals()
        self.setAcceptDrops(True)
        self.populate_table()

    def dragEnterEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
        else:
            event.ignore()

    def dropEvent(self, event):
        if event.mimeData().hasUrls():
            paths = [url.toLocalFile() for url in event.mimeData().urls()]
            add_dialog = AddJobDialog(self)
            add_dialog.populate_sources_from_paths(paths)
            if add_dialog.exec():
                new_jobs = add_dialog.get_discovered_jobs()
                if new_jobs:
                    self._logic.add_jobs(new_jobs)
            event.acceptProposedAction()
        else:
            event.ignore()

    def _build_ui(self):
        layout = QVBoxLayout(self)
        self.table = QTableWidget()
        self.table.setAcceptDrops(True)
        self.table.setSelectionMode(QAbstractItemView.ExtendedSelection)
        self.table.setSelectionBehavior(QAbstractItemView.SelectRows)
        self.table.setEditTriggers(QAbstractItemView.NoEditTriggers)
        self.table.verticalHeader().setVisible(False)
        self.table.horizontalHeader().setStretchLastSection(True)
        self.table.setContextMenuPolicy(Qt.CustomContextMenu)
        layout.addWidget(self.table)

        button_layout = QHBoxLayout()
        self.add_job_btn = QPushButton("Add Job(s)...")
        self.remove_btn = QPushButton("Remove Selected")
        self.move_up_btn = QPushButton("Move Up")
        self.move_down_btn = QPushButton("Move Down")

        button_layout.addWidget(self.add_job_btn)
        button_layout.addStretch()
        button_layout.addWidget(self.move_up_btn)
        button_layout.addWidget(self.move_down_btn)
        button_layout.addWidget(self.remove_btn)
        layout.addLayout(button_layout)

        dialog_btns = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        self.ok_button = dialog_btns.button(QDialogButtonBox.Ok)
        self.ok_button.setText("Start Processing Queue")
        layout.addWidget(dialog_btns)

        self.ok_button.clicked.connect(self.accept)
        dialog_btns.rejected.connect(self.reject)

    def _connect_signals(self):
        self.table.itemDoubleClicked.connect(
            lambda item: self._logic.configure_job_at_row(item.row())
        )
        self.table.customContextMenuRequested.connect(self._show_context_menu)
        self.add_job_btn.clicked.connect(self._logic.add_jobs_from_dialog)
        self.remove_btn.clicked.connect(self._logic.remove_selected_jobs)
        self.move_up_btn.clicked.connect(lambda: self.move_selected_jobs(-1))
        self.move_down_btn.clicked.connect(lambda: self.move_selected_jobs(1))
        QShortcut(
            QKeySequence(Qt.CTRL | Qt.Key_Up), self, lambda: self.move_selected_jobs(-1)
        )
        QShortcut(
            QKeySequence(Qt.CTRL | Qt.Key_Down),
            self,
            lambda: self.move_selected_jobs(1),
        )

    def move_selected_jobs(self, direction: int):
        selected_rows = sorted(
            [r.row() for r in self.table.selectionModel().selectedRows()]
        )
        if not selected_rows:
            return

        if direction == -1 and selected_rows[0] > 0:
            for row_index in selected_rows:
                self._logic.jobs.insert(row_index - 1, self._logic.jobs.pop(row_index))
            new_selection_start = selected_rows[0] - 1
        elif direction == 1 and selected_rows[-1] < len(self._logic.jobs) - 1:
            for row_index in reversed(selected_rows):
                self._logic.jobs.insert(row_index + 1, self._logic.jobs.pop(row_index))
            new_selection_start = selected_rows[0] + 1
        else:
            return

        self.populate_table()
        selection_model = self.table.selectionModel()
        selection_model.clearSelection()
        for i in range(len(selected_rows)):
            index = self.table.model().index(new_selection_start + i, 0)
            selection_model.select(
                index, QItemSelectionModel.Select | QItemSelectionModel.Rows
            )

    def populate_table(self):
        self.table.selectionModel().clear()
        self._logic.populate_table()

    def _show_context_menu(self, pos: Qt.Point):
        selected_rows = self.table.selectionModel().selectedRows()
        if not selected_rows:
            return

        menu = QMenu()
        config_action = menu.addAction("Configure...")
        remove_action = menu.addAction("Remove from Queue")
        menu.addSeparator()
        copy_action = menu.addAction("Copy Layout")
        paste_action = menu.addAction("Paste Layout")

        config_action.setEnabled(len(selected_rows) == 1)

        # Enable "Copy" if a single, configured job is selected
        source_job_index = selected_rows[0].row()
        source_job = self._logic.jobs[source_job_index]
        is_configured = source_job.get("status") == "Configured"
        copy_action.setEnabled(len(selected_rows) == 1 and is_configured)

        # Enable "Paste" if the clipboard has content
        paste_action.setEnabled(self._logic._layout_clipboard is not None)

        action = menu.exec(self.table.viewport().mapToGlobal(pos))

        if action == config_action:
            self._logic.configure_job_at_row(source_job_index)
        elif action == remove_action:
            self._logic.remove_selected_jobs()
        elif action == copy_action:
            self._logic.copy_layout(source_job_index)
        elif action == paste_action:
            self._logic.paste_layout()

    def get_final_jobs(self) -> list[dict]:
        return self._logic.get_final_jobs()

File: qt_ui/job_queue_dialog/ui.cpp
// Job Queue Dialog Implementation

#include "ui.hpp"
#include "logic.hpp"
#include "../add_job_dialog/ui.hpp"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QPushButton>
#include <QDialogButtonBox>
#include <QHeaderView>
#include <QMenu>
#include <QShortcut>
#include <QKeySequence>
#include <QDragEnterEvent>
#include <QDropEvent>
#include <QMimeData>
#include <QUrl>
#include <QFileInfo>

JobQueueDialog::JobQueueDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Job Queue");
    setMinimumSize(1000, 500);
    setAcceptDrops(true);

    buildUi();
    connectSignals();

    m_logic = std::make_unique<JobQueueLogic>(this);
    populateTable();
}

JobQueueDialog::~JobQueueDialog() = default;

void JobQueueDialog::buildUi()
{
    auto* mainLayout = new QVBoxLayout(this);

    // Table
    m_table = new QTableWidget();
    m_table->setAcceptDrops(true);
    m_table->setSelectionMode(QAbstractItemView::ExtendedSelection);
    m_table->setSelectionBehavior(QAbstractItemView::SelectRows);
    m_table->setEditTriggers(QAbstractItemView::NoEditTriggers);
    m_table->verticalHeader()->setVisible(false);
    m_table->horizontalHeader()->setStretchLastSection(true);
    m_table->setContextMenuPolicy(Qt::CustomContextMenu);
    mainLayout->addWidget(m_table);

    // Button row
    auto* buttonLayout = new QHBoxLayout();

    m_addJobBtn = new QPushButton("Add Job(s)...");
    buttonLayout->addWidget(m_addJobBtn);

    buttonLayout->addStretch();

    m_moveUpBtn = new QPushButton("Move Up");
    buttonLayout->addWidget(m_moveUpBtn);

    m_moveDownBtn = new QPushButton("Move Down");
    buttonLayout->addWidget(m_moveDownBtn);

    m_removeBtn = new QPushButton("Remove Selected");
    buttonLayout->addWidget(m_removeBtn);

    mainLayout->addLayout(buttonLayout);

    // Dialog buttons
    auto* dialogBtns = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);

    m_okButton = dialogBtns->button(QDialogButtonBox::Ok);
    m_okButton->setText("Start Processing Queue");

    connect(dialogBtns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(dialogBtns, &QDialogButtonBox::rejected, this, &QDialog::reject);

    mainLayout->addWidget(dialogBtns);
}

void JobQueueDialog::connectSignals()
{
    connect(m_table, &QTableWidget::cellDoubleClicked,
            this, &JobQueueDialog::onTableDoubleClicked);
    connect(m_table, &QTableWidget::customContextMenuRequested,
            this, &JobQueueDialog::onContextMenuRequested);

    connect(m_addJobBtn, &QPushButton::clicked,
            this, &JobQueueDialog::onAddJobsClicked);
    connect(m_removeBtn, &QPushButton::clicked,
            this, &JobQueueDialog::onRemoveSelectedClicked);
    connect(m_moveUpBtn, &QPushButton::clicked,
            this, &JobQueueDialog::onMoveUpClicked);
    connect(m_moveDownBtn, &QPushButton::clicked,
            this, &JobQueueDialog::onMoveDownClicked);

    // Keyboard shortcuts
    auto* moveUpShortcut = new QShortcut(QKeySequence(Qt::CTRL | Qt::Key_Up), this);
    connect(moveUpShortcut, &QShortcut::activated, this, &JobQueueDialog::onMoveUpClicked);

    auto* moveDownShortcut = new QShortcut(QKeySequence(Qt::CTRL | Qt::Key_Down), this);
    connect(moveDownShortcut, &QShortcut::activated, this, &JobQueueDialog::onMoveDownClicked);
}

void JobQueueDialog::dragEnterEvent(QDragEnterEvent* event)
{
    if (event->mimeData()->hasUrls()) {
        event->acceptProposedAction();
    } else {
        event->ignore();
    }
}

void JobQueueDialog::dropEvent(QDropEvent* event)
{
    if (event->mimeData()->hasUrls()) {
        QStringList paths;
        for (const QUrl& url : event->mimeData()->urls()) {
            paths.append(url.toLocalFile());
        }

        AddJobDialog dialog(this);
        dialog.populateSourcesFromPaths(paths);
        if (dialog.exec() == QDialog::Accepted) {
            auto discovered = dialog.getDiscoveredJobs();
            std::vector<JobData> jobs;
            for (const auto& d : discovered) {
                JobData job;
                job.sources = d;
                job.status = "Needs Configuration";
                // Get name from Source 1
                auto it = d.find("Source 1");
                if (it != d.end()) {
                    QFileInfo fi(it->second);
                    job.name = fi.baseName();
                }
                jobs.push_back(job);
            }
            addJobs(jobs);
        }
        event->acceptProposedAction();
    } else {
        event->ignore();
    }
}

void JobQueueDialog::populateTable()
{
    m_logic->populateTable();
}

void JobQueueDialog::addJobs(const std::vector<JobData>& jobs)
{
    m_logic->addJobs(jobs);
}

std::vector<JobData> JobQueueDialog::getFinalJobs() const
{
    return m_logic->getFinalJobs();
}

void JobQueueDialog::onAddJobsClicked()
{
    AddJobDialog dialog(this);
    if (dialog.exec() == QDialog::Accepted) {
        auto discovered = dialog.getDiscoveredJobs();
        std::vector<JobData> jobs;
        for (const auto& d : discovered) {
            JobData job;
            job.sources = d;
            job.status = "Needs Configuration";
            auto it = d.find("Source 1");
            if (it != d.end()) {
                QFileInfo fi(it->second);
                job.name = fi.baseName();
            }
            jobs.push_back(job);
        }
        addJobs(jobs);
    }
}

void JobQueueDialog::onRemoveSelectedClicked()
{
    m_logic->removeSelectedJobs();
}

void JobQueueDialog::onMoveUpClicked()
{
    moveSelectedJobs(-1);
}

void JobQueueDialog::onMoveDownClicked()
{
    moveSelectedJobs(1);
}

void JobQueueDialog::moveSelectedJobs(int direction)
{
    QList<int> selectedRows;
    for (const auto& index : m_table->selectionModel()->selectedRows()) {
        selectedRows.append(index.row());
    }

    if (selectedRows.isEmpty()) return;

    std::sort(selectedRows.begin(), selectedRows.end());

    m_logic->moveJobs(selectedRows, direction);
}

void JobQueueDialog::onTableDoubleClicked(int row, int column)
{
    Q_UNUSED(column);
    m_logic->configureJobAtRow(row);
}

void JobQueueDialog::onContextMenuRequested(const QPoint& pos)
{
    QList<int> selectedRows;
    for (const auto& index : m_table->selectionModel()->selectedRows()) {
        selectedRows.append(index.row());
    }

    if (selectedRows.isEmpty()) return;

    QMenu menu;
    QAction* configAction = menu.addAction("Configure...");
    QAction* removeAction = menu.addAction("Remove from Queue");
    menu.addSeparator();
    QAction* copyAction = menu.addAction("Copy Layout");
    QAction* pasteAction = menu.addAction("Paste Layout");

    configAction->setEnabled(selectedRows.size() == 1);
    copyAction->setEnabled(selectedRows.size() == 1);
    pasteAction->setEnabled(m_logic->hasClipboard());

    QAction* action = menu.exec(m_table->viewport()->mapToGlobal(pos));

    if (action == configAction) {
        m_logic->configureJobAtRow(selectedRows.first());
    } else if (action == removeAction) {
        m_logic->removeSelectedJobs();
    } else if (action == copyAction) {
        m_logic->copyLayout(selectedRows.first());
    } else if (action == pasteAction) {
        m_logic->pasteLayout();
    }
}
